nilearn/externals/skimage/random_walker_segmentation.py

```diff
@@ -0,0 +1,517 @@
+"""
+Random walker segmentation algorithm
+
+from *Random walks for image segmentation*, Leo Grady, IEEE Trans
+Pattern Anal Mach Intell. 2006 Nov;28(11):1768-83.
+
+Installing pyamg and using the 'cg_mg' mode of random_walker improves
+significantly the performance.
+"""
+
+import warnings
+import numpy as np
+from scipy import sparse, ndimage as ndi
+
+from sklearn.utils import as_float_array
+
+# executive summary for next code block: try to import umfpack from
+# scipy, but make sure not to raise a fuss if it fails since it's only
+# needed to speed up a few cases.
+# See discussions at:
+# https://groups.google.com/d/msg/scikit-image/FrM5IGP6wh4/1hp-FtVZmfcJ
+# http://stackoverflow.com/questions/13977970/ignore-exceptions-printed-to-stderr-in-del/13977992?noredirect=1#comment28386412_13977992
+try:
+    from scipy.sparse.linalg.dsolve import umfpack
+    old_del = umfpack.UmfpackContext.__del__
+
+    def new_del(self):
+        try:
+            old_del(self)
+        except AttributeError:
+            pass
+    umfpack.UmfpackContext.__del__ = new_del
+    UmfpackContext = umfpack.UmfpackContext()
+except:
+    UmfpackContext = None
+
+try:
+    from pyamg import ruge_stuben_solver
+    amg_loaded = True
+except ImportError:
+    amg_loaded = False
+from scipy.sparse.linalg import cg
+from ._rank_order import rank_order
+
+#-----------Laplacian--------------------
+
+
+def _make_graph_edges_3d(n_x, n_y, n_z):
+    """Returns a list of edges for a 3D image.
+
+    Parameters
+    ----------
+    n_x: integer
+        The size of the grid in the x direction.
+    n_y: integer
+        The size of the grid in the y direction
+    n_z: integer
+        The size of the grid in the z direction
+
+    Returns
+    -------
+    edges : (2, N) ndarray
+        with the total number of edges::
+
+            N = n_x * n_y * (nz - 1) +
+                n_x * (n_y - 1) * nz +
+                (n_x - 1) * n_y * nz
+
+        Graph edges with each column describing a node-id pair.
+    """
+    vertices = np.arange(n_x * n_y * n_z).reshape((n_x, n_y, n_z))
+    edges_deep = np.vstack((vertices[:, :, :-1].ravel(),
+                            vertices[:, :, 1:].ravel()))
+    edges_right = np.vstack((vertices[:, :-1].ravel(),
+                             vertices[:, 1:].ravel()))
+    edges_down = np.vstack((vertices[:-1].ravel(), vertices[1:].ravel()))
+    edges = np.hstack((edges_deep, edges_right, edges_down))
+    return edges
+
+
+def _compute_weights_3d(data, spacing, beta=130, eps=1.e-6,
```

Comment:
Yes, yet another parameter... I tried changing it but the effect was small because it affects low values.